File: calculations.py
# core/calculations.py

from dataclasses import dataclass

KM_PER_MILE = 1.609344
MILE_PER_KM = 1 / KM_PER_MILE

@dataclass
class Pace:
    minutes: float
    seconds: float
    unit: str  # "km" or "mile"

    @property
    def total_seconds(self) -> float:
        return self.minutes * 60 + self.seconds


def get_time_seconds(distance: float, pace_seconds: float) -> float:
    return distance * pace_seconds


def seconds_to_hms(total_seconds: float):
    total_seconds = int(round(total_seconds))
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60
    return hours, minutes, seconds


def km_to_miles(km: float) -> float:
    return km * MILE_PER_KM


def miles_to_km(miles: float) -> float:
    return miles * KM_PER_MILE


def average_speed(distance_km: float, total_seconds: float):
    hours = total_seconds / 3600.0
    if hours == 0:
        return 0.0, 0.0
    speed_kmh = distance_km / hours
    speed_mph = km_to_miles(distance_km) / hours
    return speed_kmh, speed_mph

File: calculations.java
package core;

public class Calculations {

    public static final double KM_PER_MILE = 1.609344;
    public static final double MILE_PER_KM = 1.0 / KM_PER_MILE;

    // --- Pace class (equivalent to @dataclass) ---
    public static class Pace {
        private final double minutes;
        private final double seconds;
        private final String unit; // "km" or "mile"

        public Pace(double minutes, double seconds, String unit) {
            this.minutes = minutes;
            this.seconds = seconds;
            this.unit = unit;
        }

        public double getMinutes() {
            return minutes;
        }

        public double getSeconds() {
            return seconds;
        }

        public String getUnit() {
            return unit;
        }

        public double getTotalSeconds() {
            return (minutes * 60.0) + seconds;
        }
    }

    // --- get_time_seconds ---
    public static double getTimeSeconds(double distance, double paceSeconds) {
        return distance * paceSeconds;
    }

    // --- seconds_to_hms ---
    public static int[] secondsToHms(double totalSeconds) {
        int rounded = (int) Math.round(totalSeconds);
        int hours = rounded / 3600;
        int minutes = (rounded % 3600) / 60;
        int seconds = rounded % 60;
        return new int[]{hours, minutes, seconds};
    }

    // --- km_to_miles ---
    public static double kmToMiles(double km) {
        return km * MILE_PER_KM;
    }

    // --- miles_to_km ---
    public static double milesToKm(double miles) {
        return miles * KM_PER_MILE;
    }

    // --- average_speed ---
    // Returns (km/h, mph) just like Python returns a tuple
    public static double[] averageSpeed(double distanceKm, double totalSeconds) {
        double hours = totalSeconds / 3600.0;
        if (hours == 0) {
            return new double[]{0.0, 0.0};
        }
        double speedKmh = distanceKm / hours;
        double speedMph = kmToMiles(distanceKm) / hours;
        return new double[]{speedKmh, speedMph};
    }
}
